pkg/util/proto/document.go

```diff
@@ -219,27 +226,37 @@ func (d *Definitions) parseArbitrary(s *openapi_v2.Schema, path *Path) (Schema,
 // ParseSchema creates a walkable Schema from an openapi schema. While
 // this function is public, it doesn't leak through the interface.
 func (d *Definitions) ParseSchema(s *openapi_v2.Schema, path *Path) (Schema, error) {
+	if s.GetXRef() != "" {
+		return d.parseReference(s, path)
+	}
 	objectTypes := s.GetType().GetValue()
-	if len(objectTypes) == 1 {
+	if len(objectTypes) == 0 {
+		// in the OpenAPI schema served by older k8s versions, object definitions created from structs did not include
+		// the type:object property (they only included the "properties" property), so we need to handle this case
+		if s.GetProperties() != nil {
+			return d.parseKind(s, path)
+		} else {
+			// Definition has no type and no properties. Treat it as an arbitrary value
+			// TODO: what if it has additionalProperties or patternProperties?
+			return d.parseArbitrary(s, path)
+		}
+	} else if len(objectTypes) == 1 {
 		t := objectTypes[0]
 		switch t {
 		case object:
-			return d.parseMap(s, path)
+			if s.GetProperties() != nil {
+				return d.parseKind(s, path)
+			} else {
+				return d.parseMap(s, path)
+			}
 		case array:
 			return d.parseArray(s, path)
 		}
-
-	}
-	if s.GetXRef() != "" {
-		return d.parseReference(s, path)
-	}
-	if s.GetProperties() != nil {
-		return d.parseKind(s, path)
-	}
-	if len(objectTypes) == 0 || (len(objectTypes) == 1 && objectTypes[0] == "") {
-		return d.parseArbitrary(s, path)
+		return d.parsePrimitive(s, path)
+	} else {
+		// the OpenAPI generator never generates (nor it ever did in the past) OpenAPI type definitions with multiple types
+		return nil, newSchemaError(path, "definitions with multiple types aren't supported")
```

Comment:
Maybe this should be at the beginning

Reply:
This is now the default case in a switch statement.